### DISCUSSION TITLE
replacing some other programs with libvips, need help with commands

### DISCUSSION THREAD

**Discussion by dbojan** [libvips/libvips#2558] [Q&A]
Hello.

I am considering replacing some other programs with libvips, cause I heard it is faster.

How can I:
-shave 20 pixels from left and right side, and 40 pixels from top and bottom, of the picture.
images have different dimensions
Can I use crop without knowing image dimensions?

in graphic magick would use this: 
gm.exe convert in.tif -shave 20x40 out.tif

-append image to multipage tiff, if it does not exist, create a new multipage tif:
images have different dimensions. append image, no compression needed
equivalent of: 
tiffcp.exe -c none -a in.tif out.tif

-convert multipage tiff to pdf
images have different dimensions. tried using magicksave, but pdf created is small blank page.
in gm I would use this:
gm.exe convert in.tif -units PixelsPerInch -resize 1240x -density 150x150 -compress jpeg -quality 30 out.pdf

Thank you.

**Comment by jcupitt** [comment]
Hi @dbojan,

The libvips CLI is pretty basic compared to IM/GM. It's enough to be able to call operations in the library, but that's all. This is by design: the idea is to use standard scripting languages (eg. bash / python / ruby / etc.) to do fancy stuff.

# equivalent of shave

In python (for example, but ruby is very similar) you could write:

```python
import pyvips

# crop can be streamed from source to destination, so we hint sequential access
# this gives a good speedup for many image formats 
image = pyvips.Image.new_from_file("some-filename", access="sequential")
smaller = image.crop(20, 40, image.width - 40, image.height - 80)
smaller.write_to_file("output-filename")
```

Or in bash (relatively ugly):

```
input_filename=...
output_filename=...
width=$(vipsheader -f width $input_filename)
height=$(vipsheader -f height $input_filename)
output_width=((width - 40))
output_height=((height - 80))
vips crop $input_filename $output_filename 20 40 $output_width $output_height
```

# append page to tiff

libvips supports multipage images, but can only automatically iterate over pages which are the same size. It's designed for things like OME-TIFF and DICOM where tiff pages represent image slices. It won't work well in the prepress world where TIFFs (and PDFs) often have varying page size. I'd use `tiffcp` for your example. 

# convert to PDF

libvips has quite good PDF load support, but does not support PDF save. As you say, you can write PDFs via imagemagick, but performance will be truly awful.
